### DISCUSSION TITLE
Pruning Settings

### DISCUSSION THREAD

**Discussion by andreclaro** [AcalaNetwork/Acala#2771] [0. General]
Hello, 

We just want to keep history for the last month, which in the case of Acala and Karura is about 200k blocks.

What are the correct parameters/flags to enable pruning?  `--pruning 200000` or `--state-pruning 200000 --blocks-pruning 200000`

Thanks!

**Comment by andreclaro** [comment]
Looking at the code (https://github.com/paritytech/polkadot-sdk/blob/7f7f5fa857502b6e3649081abb6b53c3512bfedb/substrate/client/cli/src/params/pruning_params.rs#L38-L39), looks like `--pruning` is equivalent to `--state-pruning`.

I started my node with `--pruning 200000`. Can I know restart the node with `--blocks-pruning 200000` to enable blocks pruning? does it take affect or do I need to delete the existing db and start syncing from genesis?

**Comment by xlc** [comment]
This is not Acala specific so best to ask at http://github.com/paritytech/polkadot-sdk or http://substrate.stackexchange.com as there will be more people able to help.

To answer your question: `--pruning` is same as `--state-pruning` and it does not enable `--block-pruning` so you will both flag to prune both state and blocks.

Pruning mode cannot be changed AFAIK so you will need to resync the node

**Reply by andreclaro** [reply]
Hi @xlc thanks! Doesn't this line https://github.com/paritytech/polkadot-sdk/blob/7f7f5fa857502b6e3649081abb6b53c3512bfedb/substrate/client/cli/src/params/pruning_params.rs#L106C51-L106C65 means that the pruning/state-pruning also does blocks-pruning? 🤔

**Reply by xlc** [reply]
The code is indeed a bit confusing. Let me explain.

There is `struct PruningParams` and it have two fields: `state_pruning` and `blocks_pruning`. Both have type `DatabasePruningMode` (ignoring the `Option`).

`DatabasePruningMode` can be converted into `PruningMode` (for state pruning) or `BlocksPruning`.

But https://github.com/paritytech/polkadot-sdk/blob/7f7f5fa857502b6e3649081abb6b53c3512bfedb/substrate/client/cli/src/params/pruning_params.rs#L62 is converting the `state_pruning` to `PruningMode`
and https://github.com/paritytech/polkadot-sdk/blob/7f7f5fa857502b6e3649081abb6b53c3512bfedb/substrate/client/cli/src/params/pruning_params.rs#L67 convert `blocks_pruning` to `BlocksPruning`

So there are two different thing just reusing some code

**Reply by andreclaro** [reply]
I have just added `--blocks-pruning 200000` to both acala and karura nodes and I see a decrease in the storage usage, and I didn't see any error or warning. 🤔 

Pruning settings now are: `--state-pruning 200000 --blocks-pruning 200000`

Acala:
![image](https://github.com/AcalaNetwork/Acala/assets/33039405/021dd4d9-52bf-4d74-baa9-016e39337758)

Karura:
![image](https://github.com/AcalaNetwork/Acala/assets/33039405/bab6794b-ac86-48c9-863a-5e81448005b7)

I just wanna make sure if I really need to delete database and start nodes from scratch.

**Reply by andreclaro** [reply]
@xlc , my doubt is: For `--blocks-pruning` to be applied by the node, do I really need to start the node from scratch (new database)? 🤔 

"This setting can only be set on the first creation of the database." -> it seems this only applies to state pruning: https://github.com/paritytech/polkadot-sdk/blob/7f7f5fa857502b6e3649081abb6b53c3512bfedb/substrate/client/cli/src/params/pruning_params.rs#L28-L37

Can you confirm? Thanks

**Reply by bkchr** [reply]
You can change blocks pruning, but it will not download pruned blocks if you switch back to archive for example. For the state pruning you can currently change the number of state to keep.

**Reply by andreclaro** [reply]
Ok! Thanks @bkchr! 

So, which settings are required to be set on the first creation of the database? Only `--state-pruning`? 🤔 

I started the nodes with `--state-pruning 200000`  and then after 2 days added `--blocks-pruning 200000`... would like to know if the node is now also pruning the blocks bodies?!?! or do I need to run the nodes from scratch with this new configuration??

**Reply by andreclaro** [reply]
The idea of using pruning is to save on storage costs, but I don't actually see a reduction of storage usage... Do I need to wait the node to be fully in sync (finalized blocks)???

The current acala node (`--pruning=archive`) is using 635G of storage.

The new acala node (initially with `--state-pruning 200000` and then I added `--blocks-pruning 200000`) is already using 484G and is still syncing the finalized blocks:

![image](https://github.com/AcalaNetwork/Acala/assets/33039405/372da436-b23c-4adb-a11d-4ea0cc6ab703)

![image](https://github.com/AcalaNetwork/Acala/assets/33039405/2f4d7aec-5743-4464-ba31-616e0eba03c4)

The huge decreases of storage utilization were due to storage capacity increase.

in the last hours:
![image](https://github.com/AcalaNetwork/Acala/assets/33039405/9a5e81de-b90a-4015-9806-ee598a43c75d)

The decrease at 9.30am was when I configured `--blocks-pruning 200000`

**Reply by andreclaro** [reply]
here are the settings I'm using:

```
/usr/local/bin/acala \
    --base-path=/var/lib/acala/data \
    --chain=acala \
    --name=NAME \
    --database=paritydb \
    --state-pruning=200000 \
    --blocks-pruning=200000 \
    --rpc-cors=all \
    --rpc-external \
    --rpc-port=9933 \
    --rpc-max-connections=500 \
    --port=30333 \
    --prometheus-external \
    --prometheus-port=9615
```

**Reply by xlc** [reply]
the main storage usage is actually coming from the relaychain side, which you did not have pruning settings.
so something like this

```
/usr/local/bin/acala \
    --base-path=/var/lib/acala/data \
    --chain=acala \
    --name=NAME \
    --database=paritydb \
    --state-pruning=200000 \
    --blocks-pruning=200000 \
    --rpc-cors=all \
    --rpc-external \
    --rpc-port=9933 \
    --rpc-max-connections=500 \
    --port=30333 \
    --prometheus-external \
    --prometheus-port=9615 \
    -- \
    --state-pruning=200000 \
    --blocks-pruning=200000
```

OR if you already have a trusted polkadot node (i.e. your own node), you can use `--relay-chain-rpc-url=wss://polkadot-rpc-url`. In this way, you can share a single polkadot node between multiple parachain nodes (that could be from different parachain networks).

**Reply by andreclaro** [reply]
@xlc, Do I need to start the node from scratch to enable that pruning settings in polkadot db ?

**Reply by xlc** [reply]
you should resync the polkadot db by deleting the polkadot folder in the base path. but can keep the acala one

**Reply by andreclaro** [reply]
Ok! Thanks! I already deleted both databases... It should take a full week to have both nodes in-sync... is there any quick way to resync the nodes with pruning settings? 🤔 

these are the settings I used (karura example):

```
/usr/local/bin/acala \
    --base-path=/var/lib/karura/data \
    --chain=karura \
    --name=xLabs \
# TODO(andre): this condition will remove when we deprecate the current nodes
# paritydb database is recommend when pruning is enabled
    --database=paritydb \
    --state-pruning=200000 \
    --blocks-pruning=200000 \
    --rpc-cors=all \
    --rpc-external \
    --rpc-port=9933 \
    --rpc-max-connections=500 \
    --port=30333 \
    --prometheus-external \
    --prometheus-port=9615 \
    -- \
    --database=paritydb \
    --state-pruning=200000 \
    --blocks-pruning=200000
```

Used `--database=paritydb` because I got the following warning message:

`acala[1810229]: 2024-06-15 10:49:42 [Relaychain] Large pruning window of 200000 detected! THIS CAN LEAD TO HIGH MEMORY USAGE AND CRASHES. Reduce the pruning window or switch your database to paritydb.`

**Reply by xlc** [reply]
I had same request and it isn't supported yet https://github.com/paritytech/polkadot-sdk/issues/8

**Reply by andreclaro** [reply]
Acala mainnet is currently stuck on finalized block #5446885 .. For details consult https://github.com/AcalaNetwork/Acala/issues/2774
